File: android-agent/app/src/main/java/com/flowlink/agent/SignalingClient.kt
package com.flowlink.agent

import com.google.gson.Gson
import okhttp3.*
import org.json.JSONObject
import java.util.concurrent.TimeUnit

class SignalingClient(private val serverUrl: String) {
    private var webSocket: WebSocket? = null
    private val client = OkHttpClient.Builder()
        .readTimeout(0, TimeUnit.MILLISECONDS)
        .build()
    private val gson = Gson()
    
    var onDeviceRegistered: ((String) -> Unit)? = null
    var onSessionRequest: ((String, String) -> Unit)? = null
    var onSessionStarted: ((String) -> Unit)? = null
    var onDisconnected: (() -> Unit)? = null
    var onOffer: ((String, String) -> Unit)? = null
    var onAnswer: ((String, String) -> Unit)? = null
    var onIceCandidate: ((String, String) -> Unit)? = null

    fun connect(deviceName: String, deviceType: String, platform: String) {
        val request = Request.Builder()
            .url(serverUrl)
            .build()

        webSocket = client.newWebSocket(request, object : WebSocketListener() {
            override fun onOpen(webSocket: WebSocket, response: Response) {
                register(deviceName, deviceType, platform)
            }

            override fun onMessage(webSocket: WebSocket, text: String) {
                handleMessage(text)
            }

            override fun onClosing(webSocket: WebSocket, code: Int, reason: String) {
                onDisconnected?.invoke()
            }

            override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
                onDisconnected?.invoke()
            }
        })
    }

    fun disconnect() {
        webSocket?.close(1000, "User disconnect")
        webSocket = null
    }

    private fun register(deviceName: String, deviceType: String, platform: String) {
        val message = JSONObject().apply {
            put("type", "register_device")
            put("timestamp", System.currentTimeMillis())
            put("device", JSONObject().apply {
                put("name", deviceName)
                put("type", deviceType)
                put("platform", platform)
            })
        }
        send(message.toString())
    }

    private fun handleMessage(text: String) {
        val json = JSONObject(text)
        val type = json.getString("type")

        when (type) {
            "device_registered" -> {
                val deviceId = json.getString("deviceId")
                onDeviceRegistered?.invoke(deviceId)
            }
            "session_request" -> {
                val sessionId = json.getString("sessionId")
                val fromDevice = json.getJSONObject("fromDevice").getString("name")
                onSessionRequest?.invoke(sessionId, fromDevice)
            }
            "session_started" -> {
                val sessionId = json.getString("sessionId")
                onSessionStarted?.invoke(sessionId)
            }
            "offer" -> {
                val sessionId = json.getString("sessionId")
                val sdp = json.getString("sdp")
                onOffer?.invoke(sessionId, sdp)
            }
            "answer" -> {
                val sessionId = json.getString("sessionId")
                val sdp = json.getString("sdp")
                onAnswer?.invoke(sessionId, sdp)
            }
            "ice_candidate" -> {
                val sessionId = json.getString("sessionId")
                val candidate = json.getJSONObject("candidate").toString()
                onIceCandidate?.invoke(sessionId, candidate)
            }
            "ping" -> {
                val pong = JSONObject().apply {
                    put("type", "pong")
                    put("timestamp", System.currentTimeMillis())
                }
                send(pong.toString())
            }
        }
    }

    fun acceptSession(sessionId: String) {
        val message = JSONObject().apply {
            put("type", "accept_session")
            put("timestamp", System.currentTimeMillis())
            put("sessionId", sessionId)
        }
        send(message.toString())
    }

    fun rejectSession(sessionId: String, reason: String) {
        val message = JSONObject().apply {
            put("type", "reject_session")
            put("timestamp", System.currentTimeMillis())
            put("sessionId", sessionId)
            put("reason", reason)
        }
        send(message.toString())
    }

    fun sendOffer(sessionId: String, sdp: String) {
        val message = JSONObject().apply {
            put("type", "offer")
            put("timestamp", System.currentTimeMillis())
            put("sessionId", sessionId)
            put("sdp", sdp)
        }
        send(message.toString())
    }

    fun sendAnswer(sessionId: String, sdp: String) {
        val message = JSONObject().apply {
            put("type", "answer")
            put("timestamp", System.currentTimeMillis())
            put("sessionId", sessionId)
            put("sdp", sdp)
        }
        send(message.toString())
    }

    fun sendIceCandidate(sessionId: String, candidate: String) {
        val message = JSONObject().apply {
            put("type", "ice_candidate")
            put("timestamp", System.currentTimeMillis())
            put("sessionId", sessionId)
            put("candidate", JSONObject(candidate))
        }
        send(message.toString())
    }

    private fun send(message: String) {
        webSocket?.send(message)
    }
}

File: desktop-agent/src/SignalingClient.ts
import WebSocket from 'ws';
import { EventEmitter } from 'events';
import { Message, MessageType, Device } from './protocol';

export class SignalingClient extends EventEmitter {
  private ws: WebSocket | null = null;
  private deviceId: string | null = null;
  private reconnectTimer: NodeJS.Timeout | null = null;

  constructor(
    private serverUrl: string,
    private deviceInfo: Omit<Device, 'id' | 'online' | 'lastSeen'>
  ) {
    super();
  }

  async connect(): Promise<void> {
    return new Promise((resolve, reject) => {
      this.ws = new WebSocket(this.serverUrl);

      this.ws.on('open', () => {
        console.log('Connected to signaling server');
        this.register();
        resolve();
      });

      this.ws.on('message', (data: Buffer) => {
        try {
          const message: Message = JSON.parse(data.toString());
          this.handleMessage(message);
        } catch (error) {
          console.error('Error parsing message:', error);
        }
      });

      this.ws.on('close', () => {
        console.log('Disconnected from signaling server');
        this.emit('disconnected');
        this.scheduleReconnect();
      });

      this.ws.on('error', (error) => {
        console.error('WebSocket error:', error);
        reject(error);
      });
    });
  }

  disconnect(): void {
    if (this.reconnectTimer) {
      clearTimeout(this.reconnectTimer);
      this.reconnectTimer = null;
    }
    
    if (this.ws) {
      this.ws.close();
      this.ws = null;
    }
  }

  private register(): void {
    this.send({
      type: MessageType.REGISTER_DEVICE,
      timestamp: Date.now(),
      device: this.deviceInfo
    });
  }

  private handleMessage(message: Message): void {
    switch (message.type) {
      case MessageType.DEVICE_REGISTERED:
        this.deviceId = (message as any).deviceId;
        this.emit('device-registered', this.deviceId);
        break;
      
      case MessageType.DEVICE_LIST:
        this.emit('device-list', (message as any).devices);
        break;
      
      case MessageType.SESSION_REQUEST:
        this.emit('session-request', {
          sessionId: (message as any).sessionId,
          fromDevice: (message as any).fromDevice
        });
        break;
      
      case MessageType.SESSION_STARTED:
        this.emit('session-started', (message as any).sessionId);
        break;
      
      case MessageType.END_SESSION:
        this.emit('session-ended', (message as any).sessionId);
        break;
      
      case MessageType.OFFER:
        this.emit('offer', {
          sessionId: (message as any).sessionId,
          sdp: (message as any).sdp
        });
        break;
      
      case MessageType.ANSWER:
        this.emit('answer', {
          sessionId: (message as any).sessionId,
          sdp: (message as any).sdp
        });
        break;
      
      case MessageType.ICE_CANDIDATE:
        this.emit('ice-candidate', {
          sessionId: (message as any).sessionId,
          candidate: (message as any).candidate
        });
        break;
      
      case MessageType.ERROR:
        this.emit('error', {
          code: (message as any).code,
          message: (message as any).message
        });
        break;
      
      case MessageType.PING:
        this.send({
          type: MessageType.PONG,
          timestamp: Date.now()
        });
        break;
    }
  }

  requestSession(targetDeviceId: string, password?: string): void {
    this.send({
      type: MessageType.REQUEST_SESSION,
      timestamp: Date.now(),
      targetDeviceId,
      password
    });
  }

  acceptSession(sessionId: string): void {
    this.send({
      type: MessageType.ACCEPT_SESSION,
      timestamp: Date.now(),
      sessionId
    });
  }

  rejectSession(sessionId: string, reason?: string): void {
    this.send({
      type: MessageType.REJECT_SESSION,
      timestamp: Date.now(),
      sessionId,
      reason
    });
  }

  endSession(sessionId: string): void {
    this.send({
      type: MessageType.END_SESSION,
      timestamp: Date.now(),
      sessionId
    });
  }

  sendOffer(sessionId: string, sdp: string): void {
    this.send({
      type: MessageType.OFFER,
      timestamp: Date.now(),
      sessionId,
      sdp
    });
  }

  sendAnswer(sessionId: string, sdp: string): void {
    this.send({
      type: MessageType.ANSWER,
      timestamp: Date.now(),
      sessionId,
      sdp
    });
  }

  sendIceCandidate(sessionId: string, candidate: RTCIceCandidateInit): void {
    this.send({
      type: MessageType.ICE_CANDIDATE,
      timestamp: Date.now(),
      sessionId,
      candidate
    });
  }

  private send(message: any): void {
    if (this.ws && this.ws.readyState === WebSocket.OPEN) {
      this.ws.send(JSON.stringify(message));
    }
  }

  private scheduleReconnect(): void {
    if (this.reconnectTimer) return;
    
    this.reconnectTimer = setTimeout(() => {
      console.log('Attempting to reconnect...');
      this.reconnectTimer = null;
      this.connect().catch(console.error);
    }, 5000);
  }
}
